howso/utilities/reaction.py

```diff
@@ -0,0 +1,294 @@
+    any existing reactions.
+
+    All individual action items (cases) can be returned with their
+    corresponding details via the ``gen_cases`` generator. The returned pair of
+    values will be returned as a Reaction.
+
+    Parameters
+    ----------
+    action : Union[pandas.DataFrame, list, dict], default None
+        (Optional) A DataFrame with columns representing the requested
+        features of ``react`` or ``react_series`` cases.
+
+    details : List or None
+        (Optional) The details of results from ``react`` or ``react_series``
+        when providing a ``details`` parameter.
+    """
+
+    __slots__ = ("_data", "_reorganized_details")
+
+    SPECIAL_KEYS = {"action_features", }
+    KNOWN_KEYS = {
+        "boundary_cases_familiarity_convictions", "boundary_cases",
+        "case_contributions", "case_feature_contributions",
+        "case_feature_residuals", "case_mda",
+        "categorical_action_probabilities", "context_values",
+        "derivation_parameters", "distance_contribution",
+        "distance_ratio_parts", "distance_ratio", "feature_contributions",
+        "feature_mda_ex_post", "feature_mda", "feature_residuals",
+        "global_case_feature_residual_convictions", "hypothetical_values",
+        "influential_cases_familiarity_convictions",
+        "influential_cases_raw_weights", "influential_cases",
+        "local_case_feature_residual_convictions", "most_similar_case_indices",
+        "most_similar_cases", "observational_errors",
+        "outlying_feature_values", "robust_influences", "robust_residuals",
+        "similarity_conviction",
+    }
+
+    def __init__(self,
+                 action: t.Optional[t.Union[pd.DataFrame, list, dict]] = None,
+                 details: t.Optional[t.MutableMapping[str, t.Any]] = None
+                 ):
+        """Initialize the dictionary with the allowed keys."""
+        self._data = {
+            'action': None,
+            'details': {}
+        }
+
+        if details is None:
+            details = {}
+
+        if action is not None:
+            self.add_reaction(action, details)
+        elif details:
+            self._data['details'] = details
+
+        self._reorganized_details = None
+
+    def _validate_key(self, key) -> str:
+        """
+        Raise KeyError if key is not one of the allowed keys.
+
+        Parameters
+        ----------
+        key : str
+            The key to check.
+
+        Returns
+        -------
+        str
+            The validated key.
+
+        Raises
+        ------
+        KeyError
+            If the given `key` is not accepted.
+        """
+        # These replacements are for convenience as we switch from "series" to
+        # "action" and "explanations" to "details".
+        if key in ("series", "explanation"):
+            warnings.warn(
+                "The keys 'series' and 'explanation' are deprecated will be "
```

Comment:
grammar here seems a little funky. "... are deprecated **and** will be removed..." may be better